### ISSUE TITLE
Cannot use injector.instanceOf() with @Cached

### ISSUE THREAD

**Issue by zhengcan** [playframework/playframework#7615]
### Are you looking for help?

No

### Play Version (2.5.x / etc)

2.6.1

### API (Scala / Java / Neither / Both)

Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Windows 7

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

### Library Dependencies

No

### Expected Behavior

Please refer the following code.
```
@Inject
play.inject.Injector injector;

@play.cache.Cached(key = "somewhat")
public Result index() {
  AnyClass o = injector.instanceOf(AnyClass.class);
  return ok();
}
```

I hope the instance of `AnyClass` will be retrived as expected.

### Actual Behavior

In fact, the injector throw an exception in DEV mode in sbt interactive mode.

```
! @74j450acf - Internal server error, for (GET) [/] ->

play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[CompletionException: java.lang.SecurityException: setContextClassLoader]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
        at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala)
        at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:204)
        at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:164)
        at play.core.j.JavaHttpErrorHandlerAdapter.$anonfun$onServerError$1(JavaHttpErrorHandlerAdapter.scala:22)
Caused by: java.util.concurrent.CompletionException: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:961)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
        at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
Caused by: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
        at play.api.inject.ContextClassLoaderInjector.withContext(Injector.scala:125)
        at play.api.inject.ContextClassLoaderInjector.instanceOf(Injector.scala:118)
        at play.inject.DelegateInjector.instanceOf(DelegateInjector.java:22)
        at controllers.Home.index(Home.java:49)
```

In PROD mode, the code works as expected.

### Reproducible Test Case

Just use the attached code as before.

**Comment by zhengcan** [comment]
As workaround, I bind CachedAction a customized implementation, which extends CachedAction by using SyncCacheApi. It works very well.

```
// In guice module
bind(CachedAction.class).to(SyncCachedAction.class);
```

```
public class SyncCachedAction extends CachedAction {

  private SyncCacheApi cacheApi;

  @Inject
  public SyncCachedAction(AsyncCacheApi asyncCacheApi, SyncCacheApi cacheApi) {
    super(asyncCacheApi);
    this.cacheApi = cacheApi;
  }

  public CompletionStage<Result> call(Http.Context ctx) {
    final String key = configuration.key();
    final Integer duration = configuration.duration();
    return CompletableFuture.supplyAsync(() ->
      cacheApi.getOrElseUpdate(key, () -> delegate.call(ctx).toCompletableFuture().get(), duration)
    );
  }

}
```

**Comment by richdougherty** [comment]
This bit stood out to me:

```
Caused by: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
```

I did some research and it looks like this may be related to a change in the security behaviour of the ForkJoinPool in later versions of Java 8. Would you mind trying with an early release of Java 8 to see if that fixes the issue for you? It looks like the change occurred in build 122, so build 121 or earlier should be OK.

References:
- Someone with the same issue: https://stackoverflow.com/questions/21440563/webstart-application-securityexception-setcontextclassloader/21460363#21460363.
- Proposal for Java 8 change on mailing list: http://cs.oswego.edu/pipermail/concurrency-interest/2013-December/012123.html
- Patch to JDK: http://hg.openjdk.java.net/jdk8/jdk8/jdk/diff/e2bdddb8bedf/src/share/classes/java/util/concurrent/ForkJoinPool.java
- Summary of changes for build 122: http://mail.openjdk.java.net/pipermail/jdk8-dev/2013-December/003860.html

**Comment by marcospereira** [comment]
@richdougherty trying with the following versions and the error still happens:

```
java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b16, mixed mode)
```
```
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)
```
I think the problem here is that we are moving between thread pools. Since `CacheAction` is not being specific about which thread pool should be used ([`thenComposeAsync` here](https://github.com/playframework/playframework/blob/master/framework/src/play-cache/src/main/java/play/cache/CachedAction.java#L30) uses a default thread pool, which is [`ForkJoinPool.commonPool`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#commonPool--)).

But, if you specify the `Executor`, then you won't need to fallback to `ForkJoinPool` which, as explained in the links you posted, is paranoid about SecurityManagers. The following works:

```java
public class MyCachedAction extends Action<MyCached> {

    private AsyncCacheApi cacheApi;
    private ExecutionContextExecutor executionContext;

    @Inject
    public MyCachedAction(AsyncCacheApi cacheApi, ExecutionContextExecutor executionContext) {
        this.cacheApi = cacheApi;
        this.executionContext = executionContext;
    }

    public CompletionStage<Result> call(Context ctx) {
        final String key = configuration.key();
        final Integer duration = configuration.duration();
        return cacheApi.<Result>get(key).thenComposeAsync(cacheResult -> {
            if (cacheResult != null) {
                return CompletableFuture.completedFuture(cacheResult);
            }
            return delegate.call(ctx).thenComposeAsync(result ->
                    cacheApi.set(key, result, duration).thenApply(ignore -> result)
            );
        }, executionContext);
    }

}
```
The only difference is that we are providing a thread pool explicitly.

**Comment by bfdes** [comment]
This sort of thing also crops up when trying to use `bindFromRequest` in forms (Play 2.6.9).
Please see https://stackoverflow.com/questions/47344249, where @marcospereira referenced this issue.

My solution was to wrap the controller action in a `CompletableFuture` and to supply an explicit `Executor`, as above.

**Comment by mkurz** [comment]
I guess this issue will finally be fixed with Play 2.8.0 - see [this comment here](https://github.com/playframework/playframework/issues/8167#issuecomment-490645264).
@bfdes Can you try with Play 2.8.0-M1 and let us know if the problem still occurs?

**Comment by bfdes** [comment]
> I guess this issue will finally be fixed with Play 2.8.0 - see [this comment here](https://github.com/playframework/playframework/issues/8167#issuecomment-490645264).
> @bfdes Can you try with Play 2.8.0-M1 and let us know if the problem still occurs?

Unfortunately not. I had performed the upgrades 2.4 -> 2.5 -> 2.6 and I'm unable to go any further.
I will see if I can recreate it in a fresh Play 2.8 project.

**Comment by mkurz** [comment]
I finally found time to write a test to confirm `SecurityException` will not be thrown anymore when using `map`, `filter`, etc. on a **parallel** stream within a Play action method. See #9594

However, when using `CompletableFuture.supplyAsync(...)` and `thenxxxAsync` methods you still have to pass an explicit ec, because Java uses the common thread pool for this method by default. There is nothing Play can do about this (it's no an issue with Play actually).

I suggest closing this issue as soon as #9594 got merged.
